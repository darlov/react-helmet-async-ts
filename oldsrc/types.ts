import { HTMLAttributes, ReactElement, ReactNode } from "react";
import { ATTRIBUTE_NAMES, TAG_NAMES, TAG_PROPERTIES } from "./constants";


export interface HelmetDatumBase<T> {
  toString(): string;

  toComponent(): T
}

export type HelmetHTMLBodyDatum = HelmetDatumBase<HTMLAttributes<HTMLBodyElement>>
export type HelmetHTMLElementDatum = HelmetDatumBase<HTMLAttributes<HTMLHtmlElement>>
export type HelmetDatum = HelmetDatumBase<ReactElement[]>
export type HelmetPriorityDatum = HelmetDatumBase<ReactElement>

export interface HelmetServerState {
  base: HelmetDatum;
  bodyAttributes: HelmetHTMLBodyDatum;
  htmlAttributes: HelmetHTMLElementDatum;
  link: HelmetDatum;
  meta: HelmetDatum;
  noscript: HelmetDatum;
  script: HelmetDatum;
  style: HelmetDatum;
  title: HelmetDatum;
  priority: HelmetPriorityDatum;
}

export interface ProviderContext {
  helmet?: HelmetServerState;
}

export type FilledContext = Required<ProviderContext>;

interface OtherElementAttributes {
  [key: string]: string | number | boolean | null | undefined;
}

export type HtmlProps = JSX.IntrinsicElements['html'];
export type BodyProps = JSX.IntrinsicElements['body'];
export type LinkProps = JSX.IntrinsicElements['link'];
export type MetaProps = JSX.IntrinsicElements['meta'];
export type BaseProps = JSX.IntrinsicElements['base'];
export type NoscriptProps = JSX.IntrinsicElements['noscript'];
export type ScriptProps = JSX.IntrinsicElements['script'];
export type StyleProps = JSX.IntrinsicElements['style'];
export type TitleProps = JSX.IntrinsicElements['title'];

export type HelmetElement =
  HtmlProps
  | BodyProps
  | LinkProps
  | MetaProps
  | BaseProps
  | NoscriptProps
  | ScriptProps
  | StyleProps
  | TitleProps;

export type HelmetNode = ReactElement<HelmetElement> | ReactElement<HelmetElement>[];


export interface HelmetTags {
  baseTag: Array<HTMLBaseElement>;
  linkTags: Array<HTMLLinkElement>;
  metaTags: Array<HTMLMetaElement>;
  noscriptTags: Array<HTMLElement>;
  scriptTags: Array<HTMLScriptElement>;
  styleTags: Array<HTMLStyleElement>;
}

export interface IHelmetProviderContext {
  setHelmet: (serverState?: HelmetServerState) => void
  helmetInstances: {
    get: () => HelmetProps[]
    add: (instance: HelmetProps) => void
    remove: (instance: HelmetProps) => void
  }
}

export interface IHelmetData {
  context: {
    helmet?: HelmetServerState;
  };

  value: IHelmetProviderContext
}


export interface HelmetProps {
  async?: boolean;
  base?: BaseProps;
  bodyAttributes?: BodyProps;
  defaultTitle?: string;
  defer?: boolean;
  encodeSpecialCharacters?: boolean;
  helmetData?: IHelmetData;
  htmlAttributes?: HtmlProps;
  onChangeClientState?: (newState: IStateType, addedTags: HelmetTags, removedTags: HelmetTags) => void;
  link?: LinkProps[];
  meta?: MetaProps[];
  noscript?: NoscriptProps[];
  script?: ScriptProps[];
  style?: StyleProps[];
  title?: string;
  titleAttributes?: TitleProps;
  titleTemplate?: string;
  prioritizeSeoTags?: boolean;
}


export interface IStateType {
  baseTag: BaseProps[],
  bodyAttributes: BodyProps,
  defer?: boolean,
  encode?: boolean,
  htmlAttributes: HtmlProps,
  linkTags: LinkProps[],
  metaTags: MetaProps[],
  noscriptTags: NoscriptProps[],
  onChangeClientState: (newState: IStateType, addedTags: HelmetTags, removedTags: HelmetTags) => void,
  scriptTags: ScriptProps[],
  styleTags: StyleProps[],
  title?: string,
  titleAttributes: TitleProps,
  prioritizeSeoTags: boolean,
}

export type TagPropsType = Pick<HelmetProps, "meta" | "link" | "noscript" | "script" | "style">
export type AttributePropsType = Pick<HelmetProps, "bodyAttributes" | "htmlAttributes" | "titleAttributes">

export type AttributeNamesMap = {
  [ATTRIBUTE_NAMES.BODY]: BodyProps,
  [ATTRIBUTE_NAMES.HTML]: HtmlProps
  [ATTRIBUTE_NAMES.TITLE]: TitleProps
}


export interface IStateTypeTagNameMap {
  [TAG_NAMES.BASE]: IStateType["baseTag"],
  [TAG_NAMES.LINK]: IStateType["linkTags"],
  [TAG_NAMES.META]: IStateType["metaTags"],
  [TAG_NAMES.NOSCRIPT]: IStateType["noscriptTags"],
  [TAG_NAMES.SCRIPT]: IStateType["scriptTags"],
  [TAG_NAMES.STYLE]: IStateType["styleTags"],
}


type InferKey<T> = T extends Partial<Record<infer K, any>> ? K : never;
type InferValue<T> = T extends Partial<Record<any, infer V>> ? V : never;

/** Returns strongly-typed entries of obj. */
export const toEntries = <T extends Partial<Record<string, any>>>(obj: T) => {
  return Object.entries(obj) as [InferKey<T>, InferValue<T>][];
};




