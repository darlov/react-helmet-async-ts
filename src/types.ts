import {ReactElement} from "react";

export type TitleProps = JSX.IntrinsicElements['title'];
export type MetaProps = JSX.IntrinsicElements['meta'];
export type StyleProps = JSX.IntrinsicElements['style'];
export type ScriptProps = JSX.IntrinsicElements['script'];
export type LinkProps = JSX.IntrinsicElements['link'];
export type NoscriptProps = JSX.IntrinsicElements['noscript'];
export type BaseProps = JSX.IntrinsicElements['base'];
export type BodyProps = JSX.IntrinsicElements['body'];
export type HtmlProps = JSX.IntrinsicElements['html'];

export type TagProps =
  TitleProps
  | MetaProps
  | StyleProps
  | ScriptProps
  | LinkProps
  | NoscriptProps
  | BaseProps
  | BodyProps
  | HtmlProps;

export interface IHelmetTags {
  titleTags?: TitleProps[],
  metaTags?: MetaProps[]
  styleTags?: StyleProps[]
  scriptTags?: ScriptProps[]
  linkTags?: LinkProps[]
  noscriptTags?: NoscriptProps[],
  baseTags?: BaseProps[]
  bodyTags?: BodyProps[]
  htmlTags?: HtmlProps[]
}

export interface IHelmetInstanceState extends IHelmetTags {
  id: number
  emptyState: boolean
}

export interface IHelmetState {
  titleTag?: TitleProps,
  baseTag?: BaseProps,
  bodyTag?: BodyProps,
  htmlTag?: HtmlProps
  metaTags: MetaProps[]
  styleTags: StyleProps[],
  scriptTags: ScriptProps[],
  linkTags: LinkProps[],
  noscriptTags: NoscriptProps[],
}

export interface IHelmetDatum<T> {
  toComponent: () => T;
  toString: () => string;
}

export interface IHelmetServerState {
  title: IHelmetDatum<ReactElement>,
  base: IHelmetDatum<ReactElement>,
  body: IHelmetDatum<Required<IHelmetState>["bodyTag"]>,
  html: IHelmetDatum<Required<IHelmetState>["htmlTag"]>
  meta: IHelmetDatum<ReactElement[]>
  style: IHelmetDatum<ReactElement[]>,
  script: IHelmetDatum<ReactElement[]>,
  link: IHelmetDatum<ReactElement[]>,
  noscript: IHelmetDatum<ReactElement[]>,
}

export interface IHelmetData {
  state?: IHelmetServerState;
}

export type OnChangeClientState = (newState: IHelmetState, addedTags: IHelmetTags, removedTags: IHelmetTags) => void;

export type ArrayElement<T> = T extends Array<infer K> ? K : never;

export type MetaAttribute = keyof Pick<MetaProps, "charSet" | "name" | "httpEquiv" | "property" | "itemProp">;
export type LinkAttribute = keyof Pick<LinkProps, "rel" | "href">;
export const primaryMetaAttributes: readonly MetaAttribute[] = ["charSet", "name", "httpEquiv", "property", "itemProp"] as const
export const primaryLinkAttributes: readonly LinkAttribute[] = ["rel", "href"] as const

export const HELMET_ATTRIBUTE = 'data-rh';

