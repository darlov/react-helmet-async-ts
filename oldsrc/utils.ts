import {TAG_NAMES, TAG_PROPERTIES, ATTRIBUTE_NAMES, SEO_PRIORITY_TAGS} from './constants';
import {
  HelmetProps,
  BaseProps,
  AttributeNamesMap,
  TagPropsType,
  IStateType,
  MetaProps,
  LinkProps,
  ScriptProps
} from "./types";

enum HELMET_PROPS {
  DEFAULT_TITLE = 'defaultTitle',
  DEFER = 'defer',
  ENCODE_SPECIAL_CHARACTERS = 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE = 'onChangeClientState',
  TITLE_TEMPLATE = 'titleTemplate',
  PRIORITIZE_SEO_TAGS = 'prioritizeSeoTags'
}

const getInnermostProperty = <T extends keyof HelmetProps>(propsList: HelmetProps[], property: T) => {
  for (let i = propsList.length - 1; i >= 0; i -= 1) {
    const props = propsList[i];

    if (Object.hasOwn(props, property)) {
      return props[property];
    }
  }

  return undefined;
};

const getTitleFromPropsList = (propsList: HelmetProps[]) => {
  let innermostTitle = getInnermostProperty(propsList, TAG_NAMES.TITLE);
  const innermostTemplate = getInnermostProperty(propsList, HELMET_PROPS.TITLE_TEMPLATE);
  if (Array.isArray(innermostTitle)) {
    innermostTitle = innermostTitle.join('');
  }
  if (innermostTemplate && innermostTitle) {
    // use function arg to avoid need to escape $ characters
    return innermostTemplate.replace(/%s/g, innermostTitle);
  }

  const innermostDefaultTitle = getInnermostProperty(propsList, HELMET_PROPS.DEFAULT_TITLE);

  return innermostTitle || innermostDefaultTitle || undefined;
};

const getOnChangeClientState = (propsList: HelmetProps[]) =>
  getInnermostProperty(propsList, HELMET_PROPS.ON_CHANGE_CLIENT_STATE) || (() => {
  });

const getAttributesFromPropsList = <T extends ATTRIBUTE_NAMES>(tagType: T, propsList: HelmetProps[]) =>
  propsList
    .map(props => props[tagType])
    .filter((tag): tag is AttributeNamesMap[T] => !!tag)
    .reduce<AttributeNamesMap[T]>((tagAttrs, current) => ({...tagAttrs, ...current}), {});

const getBaseTagFromPropsList = (primaryAttributes: TAG_PROPERTIES[], propsList: HelmetProps[]) =>
  propsList
    .map(props => props.base)
    .filter((tag): tag is BaseProps => !!tag)
    .reverse()
    .reduce<BaseProps[]>((innermostBaseTag, tag) => {
      if (!innermostBaseTag.length) {
        const keys = Object.keys(tag);

        for (let i = 0; i < keys.length; i += 1) {
          const attributeKey = keys[i];
          const lowerCaseAttributeKey = attributeKey.toLowerCase();

          if (primaryAttributes.some(m => m === lowerCaseAttributeKey) && tag[lowerCaseAttributeKey as keyof BaseProps]) {
            return innermostBaseTag.concat(tag);
          }
        }
      }

      return innermostBaseTag;
    }, []);

// eslint-disable-next-line no-console
const warn = (msg: string) => console && typeof console.warn === 'function' && console.warn(msg);

type Flatten<T> = T extends any[] ? T[number] : T;

const getTagsFromPropsList = <T extends keyof TagPropsType,
  TValidTags extends Required<TagPropsType>,
  TTags = TagPropsType[T],
  TElement = Flatten<TTags>>(tagName: T, primaryAttributes: TAG_PROPERTIES[], propsList: TagPropsType[]) => {
  // Calculate list of tags, giving priority innermost component (end of the propslist)
  const approvedSeenTags: { [key in keyof TElement]: any; } = {} as TElement;

  return propsList
    .filter((props): props is TValidTags => {
      if (Array.isArray(props[tagName])) {
        return true;
      }
      if (typeof props[tagName] !== 'undefined') {
        warn(
          `Helmet: ${ tagName } should be of type "Array". Instead found type "${ typeof props[
            tagName
            ] }"`
        );
      }
      return false;
    })
    .map((props) => props[tagName])
    .reverse()
    .reduce((approvedTags, instanceTags) => {
      const instanceSeenTags: { [key in keyof TElement]: any; } = {} as TElement;
      
      instanceTags
        .filter(tag => {
          let primaryAttributeKey: string | undefined = undefined;
          const keys = Object.keys(tag);
          for (let i = 0; i < keys.length; i += 1) {
            const attributeKey = keys[i];
            const lowerCaseAttributeKey = attributeKey.toLowerCase();

            // Special rule with link tags, since rel and href are both primary tags, rel takes priority
            if (
              primaryAttributes.some(m => m === lowerCaseAttributeKey) &&
              !(
                primaryAttributeKey === TAG_PROPERTIES.REL &&
                tag[primaryAttributeKey as keyof typeof tag].toLowerCase() === 'canonical'
              ) &&
              !(
                lowerCaseAttributeKey === TAG_PROPERTIES.REL &&
                tag[lowerCaseAttributeKey as keyof typeof tag].toLowerCase() === 'stylesheet'
              )
            ) {
              primaryAttributeKey = lowerCaseAttributeKey;
            }
            // Special case for innerHTML which doesn't work lowercased
            if (
              primaryAttributes.some(m => m === attributeKey) &&
              (attributeKey === TAG_PROPERTIES.INNER_HTML ||
                attributeKey === TAG_PROPERTIES.CSS_TEXT ||
                attributeKey === TAG_PROPERTIES.ITEM_PROP)
            ) {
              primaryAttributeKey = attributeKey;
            }
          }

          if (!primaryAttributeKey || !tag[primaryAttributeKey as keyof typeof tag]) {
            return false;
          }

          const value: string = tag[primaryAttributeKey as keyof typeof tag].toLowerCase();

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement]) {
            approvedSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!instanceSeenTags[primaryAttributeKey as keyof TElement]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement] = {};
          }

          if (!approvedSeenTags[primaryAttributeKey as keyof TElement][value]) {
            instanceSeenTags[primaryAttributeKey as keyof TElement][value] = true;
            return true;
          }

          return false;
        }, [])
        .reverse()
        .forEach(tag => approvedTags.push(tag as TElement));

      // Update seen tags with tags from this instance
      const keys = Object.keys(instanceSeenTags);
      for (let i = 0; i < keys.length; i += 1) {
        const attributeKey = keys[i];
        const tagUnion = {
          ...approvedSeenTags[attributeKey as keyof TElement],
          ...instanceSeenTags[attributeKey as keyof TElement],
        };

        approvedSeenTags[attributeKey as keyof TElement] = tagUnion;
      }

      return approvedTags;
    }, [] as TElement[])
    .reverse();
};

const getAnyTrueFromPropsList = <T extends keyof HelmetProps>(propsList: HelmetProps[], checkedTag: T) => {
  if (Array.isArray(propsList) && propsList.length) {
    for (let index = 0; index < propsList.length; index += 1) {
      const prop = propsList[index];
      if (prop[checkedTag]) {
        return true;
      }
    }
  }
  return false;
};
const reducePropsToState = (propsList: HelmetProps[]): IStateType => ({
  baseTag: getBaseTagFromPropsList([TAG_PROPERTIES.HREF], propsList),
  bodyAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.BODY, propsList),
  defer: getInnermostProperty(propsList, HELMET_PROPS.DEFER),
  encode: getInnermostProperty(propsList, HELMET_PROPS.ENCODE_SPECIAL_CHARACTERS),
  htmlAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.HTML, propsList),
  linkTags: getTagsFromPropsList(
    TAG_NAMES.LINK,
    [TAG_PROPERTIES.REL, TAG_PROPERTIES.HREF],
    propsList
  ),
  metaTags: getTagsFromPropsList(
    TAG_NAMES.META,
    [
      TAG_PROPERTIES.NAME,
      TAG_PROPERTIES.CHARSET,
      TAG_PROPERTIES.HTTPEQUIV,
      TAG_PROPERTIES.PROPERTY,
      TAG_PROPERTIES.ITEM_PROP,
    ],
    propsList
  ),
  noscriptTags: getTagsFromPropsList(TAG_NAMES.NOSCRIPT, [TAG_PROPERTIES.INNER_HTML], propsList),
  onChangeClientState: getOnChangeClientState(propsList),
  scriptTags: getTagsFromPropsList(
    TAG_NAMES.SCRIPT,
    [TAG_PROPERTIES.SRC, TAG_PROPERTIES.INNER_HTML],
    propsList
  ),
  styleTags: getTagsFromPropsList(TAG_NAMES.STYLE, [TAG_PROPERTIES.CSS_TEXT], propsList),
  title: getTitleFromPropsList(propsList),
  titleAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.TITLE, propsList),
  prioritizeSeoTags: getAnyTrueFromPropsList(propsList, HELMET_PROPS.PRIORITIZE_SEO_TAGS),
});

export const flattenArray = (possibleArray: any) =>
  Array.isArray(possibleArray) ? possibleArray.join('') : possibleArray;

export { reducePropsToState };

const checkIfPropsMatch = (props: any, toMatch: any) => {
  const keys = Object.keys(props);
  for (let i = 0; i < keys.length; i += 1) {
    // e.g. if rel exists in the list of allowed props [amphtml, alternate, etc]
    if (toMatch[keys[i]] && toMatch[keys[i]].includes(props[keys[i]])) {
      return true;
    }
  }
  return false;
};

export type Prioritizer =
  { element: MetaProps, propsToMatch: typeof SEO_PRIORITY_TAGS["meta"] }
  | { element: LinkProps, propsToMatch: typeof SEO_PRIORITY_TAGS["link"] }
  | { element: ScriptProps, propsToMatch: typeof SEO_PRIORITY_TAGS["script"] };

export const prioritizer = <TElement extends Prioritizer["element"], TMatch extends Prioritizer["propsToMatch"]>(
  elements: TElement[],
  propsToMatch: TMatch) => {

  const result: { priority: TElement[], default: TElement[] } = {priority: [], default: []};

  if (Array.isArray(elements)) {

    for (const element of elements) {
      if (checkIfPropsMatch(element, propsToMatch)) {
        result.priority.push(element);
      } else {
        result.default.push(element);
      }
    }
  }
  return  result;
};

export const without = (obj: any, key: any) => {
  return {
    ...obj,
    [key]: undefined,
  };
}
