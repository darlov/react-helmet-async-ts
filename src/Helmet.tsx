import {FC, ReactNode, useEffect, useId, useMemo} from "react";
import {
  IHelmetInstanceState,
} from "./types";
import {useHelmetContext} from "./HelmetProvider";
import {Base, Title} from "./tags";
import {createActionsData, HelmetScopedContext, IHelmetScopedContextData} from "./HelmetScopedProvider";

interface IHelmetProps {
  defaultTitle?: ReactNode,
  children?: ReactNode
}

export const Helmet: FC<IHelmetProps> = ({children, defaultTitle}) => {
  const rootContext = useHelmetContext();
  const sourceId = useId();
  const id = useMemo(() => {
    return parseInt(sourceId.replace(":r", "").replace(":", ""), 32);
  }, [sourceId])

  const instanceState = useMemo<IHelmetInstanceState>(() => {
    return {
      id,
      emptyState: false
    }
  }, [id])

  useEffect(() => {
    rootContext.addInstance(instanceState);
    return () => rootContext.removeInstance(instanceState);
  }, [instanceState, rootContext.addInstance, rootContext.removeInstance])

  if (!rootContext.canUseDOM) {
    rootContext.addInstance(instanceState);
  }

  const context = useMemo<IHelmetScopedContextData>(() => {
    return {
      baseActions: createActionsData(instanceState, "baseTags", rootContext),
      htmlActions: createActionsData(instanceState, "htmlTags", rootContext),
      bodyActions: createActionsData(instanceState, "bodyTags", rootContext),
      scriptActions: createActionsData(instanceState, "scriptTags", rootContext),
      noscriptActions: createActionsData(instanceState, "noscriptTags", rootContext),
      styleActions: createActionsData(instanceState, "styleTags", rootContext),
      titleActions: createActionsData(instanceState, "titleTags", rootContext),
      metaActions: createActionsData(instanceState, "metaTags", rootContext),
      linkActions: createActionsData(instanceState, "linkTags", rootContext)      
    }
  }, [instanceState, rootContext.addItem, rootContext.removeItem])

  return (
    <HelmetScopedContext.Provider value={context}>
      {defaultTitle && <Title>{defaultTitle}</Title>}
      {children}
    </HelmetScopedContext.Provider>);
}

Helmet.displayName = "Helmet";