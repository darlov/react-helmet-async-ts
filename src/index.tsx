export { HelmetContextProvider, useHelmetContext } from "./HelmetProvider"
export { Helmet } from "./Helmet"
export * from "./tags"
export { HELMET_ATTRIBUTE } from "./Types"