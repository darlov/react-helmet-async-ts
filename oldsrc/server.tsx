import React, {ClassAttributes, createElement, DetailedHTMLProps, HTMLAttributes, ReactElement} from 'react';
import {
  ATTRIBUTE_NAMES,
  HELMET_ATTRIBUTE,
  REACT_TAG_MAP,
  SEO_PRIORITY_TAGS,
  TAG_NAMES,
  TAG_PROPERTIES,
} from './constants';
import {flattenArray, prioritizer} from './utils';
import {
  HelmetDatum,
  HelmetHTMLBodyDatum,
  HelmetHTMLElementDatum,
  HelmetServerState,
  IStateType,
  TitleProps,
  toEntries
} from "./types";

const SELF_CLOSING_TAGS = [TAG_NAMES.NOSCRIPT, TAG_NAMES.SCRIPT, TAG_NAMES.STYLE];


const encodeSpecialCharacters = (str: string, encode = true) => {
  if (!encode) {
    return String(str);
  }

  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
};

const generateElementAttributesAsString = (data: TypedTagsAttributes) =>
  toEntries(data.attributes).reduce((str, [key, value]) => {
    const attr = value !== undefined
      ? `${key}="${value}"`
      : `${key}`;
    return str ? `${str} ${attr}` : attr;
  }, '');

const generateTitleAsString = (data: TypedTagsAttrTitle, encode?: boolean) => {
  const attributeString = generateElementAttributesAsString(data);
  const flattenedTitle = flattenArray(data.title);
  return attributeString
    ? `<${data.type} ${HELMET_ATTRIBUTE}="true" ${attributeString}>${encodeSpecialCharacters(
      flattenedTitle,
      encode
    )}</${data.type}>`
    : `<${data.type} ${HELMET_ATTRIBUTE}="true">${encodeSpecialCharacters(
      flattenedTitle,
      encode
    )}</${data.type}>`;
};

const generateTagsAsString = (data: Exclude<TagType, TypedTagsAttributes>, encode?: boolean): string =>
  (data.tags as any[]).reduce((str, tag) => {
    const attributeHtml = Object.keys(tag)
      .filter(
        attribute =>
          !(attribute === TAG_PROPERTIES.INNER_HTML || attribute === TAG_PROPERTIES.CSS_TEXT)
      )
      .reduce((string, attribute) => {
        const attr =
          typeof tag[attribute] === 'undefined'
            ? attribute
            : `${attribute}="${encodeSpecialCharacters(tag[attribute], encode)}"`;
        return string ? `${string} ${attr}` : attr;
      }, '');

    const tagContent = tag.innerHTML || tag.cssText || '';

    const isSelfClosing = SELF_CLOSING_TAGS.indexOf(data.type) === -1;

    return `${str}<${data.type} ${HELMET_ATTRIBUTE}="true" ${attributeHtml}${
      isSelfClosing ? `/>` : `>${tagContent}</${data.type}>`
    }`;
  }, '');

type extractGeneric<Type> = Type extends DetailedHTMLProps<HTMLAttributes<infer T>, any> ? T : never

const convertElementAttributesToReactProps = <P extends HTMLAttributes<T>, T extends HTMLElement>(attributes: ClassAttributes<T> & P, initProps: any = {}): HTMLAttributes<T> =>
  toEntries(attributes).reduce((obj, [key, value]) => {
    obj[REACT_TAG_MAP[key as keyof typeof REACT_TAG_MAP] || key] = value;
    return obj;
  }, initProps);

const generateTitleAsReactComponent = (title: string | undefined, attributes: TitleProps) => {
  // assigning into an array to define toString function on it
  const initProps = {
    key: title,
    [HELMET_ATTRIBUTE]: true,
  };
  const props = convertElementAttributesToReactProps(attributes, initProps);

  return [createElement(TAG_NAMES.TITLE, props, title)];
};

const generateTagsAsReactComponent = <T extends Exclude<TagType, TypedTagsAttributes>>(data: T): ReactElement[] => {
  return data.tags.map((tag, i) => {
    const mappedTag: any = {
      key: i,
      [HELMET_ATTRIBUTE]: true,
    };

    toEntries(tag).forEach(([attribute, value]) => {
      if (attribute  as any == TAG_PROPERTIES.INNER_HTML || attribute as any === TAG_PROPERTIES.CSS_TEXT) {
        const content = (tag as any).innerHTML || (tag as any).cssText;
        mappedTag.dangerouslySetInnerHTML = {__html: content};
      } else {
        const mappedAttribute = REACT_TAG_MAP[attribute as unknown as keyof typeof REACT_TAG_MAP] || attribute;
        mappedTag[mappedAttribute] = value;
      }
    });

    return createElement(data.type, mappedTag);
  })
}


type TypedItem<T> = { type: T };
type TypedTagsAttr<T, TKey extends keyof IStateType> = TypedItem<T> & { attributes: IStateType[TKey] };
type TypedTags<T, TKey extends keyof IStateType> = TypedItem<T> & { tags: IStateType[TKey] };

type TypedTagsAttrTitle = TypedTagsAttr<TAG_NAMES.TITLE, "titleAttributes"> & { title: IStateType["title"] };
type TypedTagsAttrBody = TypedTagsAttr<ATTRIBUTE_NAMES.BODY, "bodyAttributes">;
type TypedTagsAttrHtml = TypedTagsAttr<ATTRIBUTE_NAMES.HTML, "htmlAttributes">;

type TypedTagsAttributes = TypedTagsAttrTitle | TypedTagsAttrBody | TypedTagsAttrHtml;

type TagType =
  TypedTagsAttributes
  | TypedTags<TAG_NAMES.STYLE, "styleTags">
  | TypedTags<TAG_NAMES.SCRIPT, "scriptTags">
  | TypedTags<TAG_NAMES.NOSCRIPT, "noscriptTags">
  | TypedTags<TAG_NAMES.META, "metaTags">
  | TypedTags<TAG_NAMES.LINK, "linkTags">
  | TypedTags<TAG_NAMES.BASE, "baseTag">;


type TagOrAttribute<T extends TagType> =
  T extends { type: ATTRIBUTE_NAMES.HTML } ? HelmetHTMLElementDatum :
    T extends { type: ATTRIBUTE_NAMES.BODY } ? HelmetHTMLBodyDatum :
      HelmetDatum;

const getMethodsForTag = <T extends TagType, TReturn extends TagOrAttribute<T>>(data: T, encode?: boolean): TagOrAttribute<T> => {
  switch (data.type) {
    case TAG_NAMES.TITLE:
      return {
        toComponent: () => generateTitleAsReactComponent(data.title, data.attributes),
        toString: () => generateTitleAsString(data, encode),
      } as any;
    case ATTRIBUTE_NAMES.BODY:
      return {
        toComponent: () => convertElementAttributesToReactProps(data.attributes),
        toString: () => generateElementAttributesAsString(data),
      } as any;
    case ATTRIBUTE_NAMES.HTML:
      return {
        toComponent: () => convertElementAttributesToReactProps(data.attributes),
        toString: () => generateElementAttributesAsString(data),
      } as any;
    default:
      return {
        toComponent: () => generateTagsAsReactComponent(data),
        toString: () => generateTagsAsString(data, encode),
      } as any;
  }
};

const getPriorityMethods = ({metaTags, linkTags, scriptTags, encode}: IStateType) => {
  const meta = prioritizer(metaTags, SEO_PRIORITY_TAGS.meta);
  const link = prioritizer(linkTags, SEO_PRIORITY_TAGS.link);
  const script = prioritizer(scriptTags, SEO_PRIORITY_TAGS.script);

  // need to have toComponent() and toString()
  const priorityMethods = {
    toComponent: () => [
      ...generateTagsAsReactComponent({type: TAG_NAMES.META, tags: meta.priority}),
      ...generateTagsAsReactComponent({type: TAG_NAMES.LINK, tags: link.priority}),
      ...generateTagsAsReactComponent({type: TAG_NAMES.SCRIPT, tags: script.priority}),
    ][0],
    toString: () =>
      // generate all the tags as strings and concatenate them
      `${getMethodsForTag({type: TAG_NAMES.META, tags: meta.priority}, encode)} ${getMethodsForTag(
        {type: TAG_NAMES.LINK, tags: link.priority},
        encode
      )} ${getMethodsForTag({type: TAG_NAMES.SCRIPT, tags: script.priority}, encode)}`,
  };

  return {
    priorityMethods,
    metaTags: meta.default,
    linkTags: link.default,
    scriptTags: script.default,
  };
};

const mapStateOnServer = (props: IStateType): HelmetServerState => {
  const {
    baseTag,
    bodyAttributes,
    encode,
    htmlAttributes,
    noscriptTags,
    styleTags,
    title = '',
    titleAttributes,
    prioritizeSeoTags,
  } = props;
  let {linkTags, metaTags, scriptTags} = props;
  let priorityMethods = {
    toComponent: (): ReactElement => (<></>),
    toString: () => '',
  };
  if (prioritizeSeoTags) {
    ({priorityMethods, linkTags, metaTags, scriptTags} = getPriorityMethods(props));
  }
  return {
    priority: priorityMethods,
    base: getMethodsForTag({type: TAG_NAMES.BASE, tags: baseTag}, encode),
    bodyAttributes: getMethodsForTag({type: ATTRIBUTE_NAMES.BODY, attributes: bodyAttributes}, encode),
    htmlAttributes: getMethodsForTag({type: ATTRIBUTE_NAMES.HTML, attributes: htmlAttributes}, encode),
    link: getMethodsForTag({type: TAG_NAMES.LINK, tags: linkTags}, encode),
    meta: getMethodsForTag({type: TAG_NAMES.META, tags: metaTags}, encode),
    noscript: getMethodsForTag({type: TAG_NAMES.NOSCRIPT, tags: noscriptTags}, encode),
    script: getMethodsForTag({type: TAG_NAMES.SCRIPT, tags: scriptTags}, encode),
    style: getMethodsForTag({type: TAG_NAMES.STYLE, tags: styleTags}, encode),
    title: getMethodsForTag({type: TAG_NAMES.TITLE, title, attributes: titleAttributes}, encode),
  };
};

export default mapStateOnServer;
