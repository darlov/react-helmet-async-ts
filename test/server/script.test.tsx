import {Helmet, Script} from '../../src';
import {render} from './utils';
import {IHelmetDataContext} from "../../src/types";
import {renderToStaticMarkup} from "react-dom/server";


describe('server', () => {
  describe('Declarative API', () => {
    it('renders script tags as React components', () => {
      const context: IHelmetDataContext = {};
      render(
        <Helmet>
          <Script src="http://localhost/test.js" type="text/javascript"/>
          <Script src="http://localhost/test2.js" type="text/javascript"/>
        </Helmet>,
        context
      );

      expect(context.state).toBeDefined();
      const {script} = context.state!;

      expect(script).toBeDefined();
      expect(script.toComponent).toBeDefined();

      const scriptComponent = script.toComponent();

      expect(scriptComponent).toHaveLength(2);

      scriptComponent.forEach(script => {
        expect(script).toEqual(expect.objectContaining({type: 'script'}));
      });

      const markup = renderToStaticMarkup(<>{scriptComponent}</>);

      expect(markup).toMatchSnapshot();
    });

    it('renders script tags as string', () => {
      const context: IHelmetDataContext = {};
      render(
        <Helmet>
          <Script src="http://localhost/test.js" type="text/javascript"/>
          <Script src="http://localhost/test2.js" type="text/javascript"/>
        </Helmet>,
        context
      );

      expect(context.state).toBeDefined();
      const {script} = context.state!;

      expect(script).toBeDefined();
      expect(script.toString).toBeDefined();
      expect(script.toString()).toMatchSnapshot();
    });
  });
});
