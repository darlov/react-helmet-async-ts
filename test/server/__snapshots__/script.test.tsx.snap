// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`server > Declarative API > renders script tags as React components 1`] = `"<script src="http://localhost/test.js" type="text/javascript" data-rh="true"></script><script src="http://localhost/test2.js" type="text/javascript" data-rh="true"></script>"`;

exports[`server > Declarative API > renders script tags as string 1`] = `"<script src="http://localhost/test.js" type="text/javascript" data-rh="true"></script><script src="http://localhost/test2.js" type="text/javascript" data-rh="true"></script>"`;
